### PR TITLE
Support vault cacert bytes env

### PR DESCRIPTION
**Pull request by tomhjp**
This builds on #446 to support setting CA certificates for injected pods, but uses `VAULT_CACERT_BYTES` instead of writing files. It turned out Vault Agent's auto-auth already supported that env var but templating did not because consul-template hadn't implemented it yet. This was fixed in Vault 1.14.2 and 1.15.0, via https://github.com/hashicorp/vault/pull/22322. Vault 1.15.0 also added the `-dev-tls-san` flag via https://github.com/hashicorp/vault/pull/22657 which lets us easily test with self-signed certificates.

To test and teardown locally, use `make image deploy exercise teardown` in a local `kind` cluster. I added that workflow to CI as well to make sure any breakages are found quickly.

### CODE REVIEW

**Review by kschoche — APPROVED** [review]
Everything looks great, I had just a question around whether or not we should require b64 encoding of the field?

**Comment by tvoran** [comment]
I think vault has a [dev-tls mode](https://developer.hashicorp.com/vault/tutorials/getting-started/getting-started-dev-server#starting-the-dev-server) now that generates its own certs, might able to use that instead of cert-manager, if we can copy the generated cert out of the vault container.

**Comment by tomhjp** [comment]
> I think vault has a [dev-tls mode](https://developer.hashicorp.com/vault/tutorials/getting-started/getting-started-dev-server#starting-the-dev-server) now that generates its own certs, might able to use that instead of cert-manager

I considered this before but didn't give it a fair shot because of the fact it would require a multi-stage Vault helm chart rollout. One part to get Vault and its certificate, and then another part to give those certificate details to the injector. I had a more serious go anyway in c8fc7ff5c6dfd3e12a25d47c192cff7f7de4f8fa, and it's actually a really nice simplification overall. Unfortunately though, `-dev-tls` doesn't support custom DNS names, so we end up with this error:

> Error checking seal status: Get "https://vault.default.svc:8200/v1/sys/seal-status": tls: failed to verify certificate: x509: certificate is valid for localhost, localhost4, localhost6, localhost.localdomain, vault-0, not vault.default.svc

We can get around it by setting `VAULT_SKIP_VERIFY` to true, but then that doesn't test the custom CA either. I might raise a separate Vault PR to support custom DNS names for `-dev-tls`, as that would be a nice option in the future.

File: agent-inject/agent/container_env.go
Comment (tomhjp) [reply agent-inject/agent/container_env.go:146]: I don't see the benefit of making b64 _required_. Supporting the raw PEM is nice because it's how the environment variable is eventually consumed anyway, and supporting b64 is nice because as you say it can be frustrating when there are multiple levels of interpretation in the automation that gets the value into place.

File: agent-inject/agent/agent.go
Comment (tomhjp) [reply agent-inject/agent/agent.go:254]: It can be PEM-encoded or base64. I added some comments in https://github.com/hashicorp/vault-k8s/pull/507/commits/d50e787e0e4cb65f43ef44bb8498d69585a2ea95 to document the optional base64 encoding.

**Comment by tomhjp** [comment]
Vault v1.14.2 should have full support for `VAULT_CACERT_BYTES` as long as https://github.com/hashicorp/vault/pull/22338 lands in time. I'm planning to hold off merging this until that releases, because otherwise we'll be stuck with the hack in this PR for a long time for backwards compatibility reasons.

**Comment by M0NsTeRRR** [comment]
Hello,
Thanks for your works :), any news to get this merged @tomhjp ?
Regards,

**Comment by tomhjp** [comment]
Thanks for the reminder @M0NsTeRRR - I just pushed a few updates that take advantage of Vault updates in 1.14.2 and 1.15.0. There should be a release of both Vault and vault-helm pretty soon, at which point I can remove the version pinning from dev.values.yaml and this should be ready to land.

@tvoran @kschoche, the updates are non-trivial so you may want to have another scan, but nothing we haven't already discussed in the PR feedback. The only remaining change I'm planning to make before merging is to get rid of the version pinning in dev.values.yaml.

**Comment by tomhjp** [comment]
Actually, we'll also be releasing vault-k8s soon, and it would be nice to get this change into that release, so perhaps I won't wait for the vault-helm release, but instead just give @tvoran and @kschoche a chance to have another pass at reviewing. I'll back out the version pinning in a separate PR after all the releases are done.

**Review by tvoran — APPROVED** [review]
Worked great for me testing locally. Would you mind updating the PR description since I think it's changed a bit since its original incarnation? And a changelog entry as well.

**Comment by kschoche** [comment]
(still) Looks good to me, I like the approach! 👍🏽

**Comment by tomhjp** [comment]
Thanks for the reviews!

> Would you mind updating the PR description since I think it's changed a bit since its original incarnation? And a changelog entry as well.

Thanks for the reminder, done 👍

**Pull request merged**
